File: tests/test_collider.cpp
#include "../maglit/collider.h"
#include <filesystem>
#include <fstream>
#include <gtest/gtest.h>
#include <string>

// Test fixture for collider functionality
class ColliderTest : public ::testing::Test {
  protected:
    void SetUp() override {
        // Create temporary directory for test files
        test_dir = "test_input_files";
        std::filesystem::create_directories(test_dir);
        // Create shape files for collider tests
        square_shape_file = test_dir + "/square_shape.txt";
        circle_shape_file = test_dir + "/circle_shape.txt";
        invalid_shape_file = test_dir + "/invalid_shape.txt";
        reversed_square_shape_file = test_dir + "/reversed_square_shape.txt";

        createShapeFiles();
    }

    void TearDown() override {
        // Clean up test files
        std::filesystem::remove_all(test_dir);
    }

    std::string test_dir;
    std::string square_shape_file;
    std::string circle_shape_file;
    std::string invalid_shape_file;
    std::string reversed_square_shape_file;

    void createShapeFiles() {
        // Create a square shape (0 to 1 in both R and Z)
        std::ofstream square(square_shape_file);
        square << "0.0 0.0\n";
        square << "1.0 0.0\n";
        square << "1.0 1.0\n";
        square << "0.0 1.0\n";
        square.close();

        // Create an approximate circle (octagon) centered at (0.5, 0.5) with radius 0.4
        std::ofstream circle(circle_shape_file);
        double        cx = 0.5, cz = 0.5, radius = 0.4;
        for (int i = 0; i < 8; i++) {
            double angle = 2.0 * M_PI * i / 8.0;
            double r = cx + radius * cos(angle);
            double z = cz + radius * sin(angle);
            circle << r << " " << z << "\n";
        }
        circle.close();

        // Create invalid shape file
        std::ofstream invalid(invalid_shape_file);
        invalid << "0.0 0.0\n";
        invalid << "1.0 0.0\n";
        // Missing third vertex
        invalid.close();

        // Create reversed winding order square shape
        std::ofstream reversed(reversed_square_shape_file);
        reversed << "0.0 1.0\n";
        reversed << "1.0 0.0\n";
        reversed << "1.0 1.0\n";
        reversed << "0.0 0.0\n";
        reversed.close();
    }
};

// ==================== COLLIDER TESTS ====================

// Test: Default constructor
TEST_F(ColliderTest, Collider_DefaultConstructor) {
    EXPECT_NO_THROW(collider shape);
}

// Test: Load valid square shape
TEST_F(ColliderTest, Collider_LoadSquareShape) {
    collider square_shape;
    square_shape.load_shape(square_shape_file);
    EXPECT_TRUE(square_shape.is_loaded());
    EXPECT_EQ(square_shape.get_vertices().size(), 4);
}

// Test: Load valid circle shape
TEST_F(ColliderTest, Collider_LoadCircleShape) {
    collider circle_shape;
    circle_shape.load_shape(circle_shape_file);
    EXPECT_TRUE(circle_shape.is_loaded());
    EXPECT_EQ(circle_shape.get_vertices().size(), 8);
}

// Test: Load invalid shape
TEST_F(ColliderTest, Collider_LoadInvalidShape) {
    collider invalid_shape;
    invalid_shape.load_shape(invalid_shape_file);
    EXPECT_FALSE(invalid_shape.is_loaded());
}

// Test: Point inside square shape
TEST_F(ColliderTest, Collider_PointInsideSquare) {
    collider square_shape;
    square_shape.load_shape(square_shape_file);
    EXPECT_TRUE(square_shape.inside(0.5, 0.5));  // Inside
    EXPECT_FALSE(square_shape.inside(1.5, 0.5)); // Outside
    EXPECT_TRUE(square_shape.inside(0.0, 0.5));  // On edge
}

// Test: Point inside circle shape
TEST_F(ColliderTest, Collider_PointInsideCircle) {
    collider circle_shape;
    circle_shape.load_shape(circle_shape_file);
    EXPECT_TRUE(circle_shape.inside(0.5, 0.5));  // Inside
    EXPECT_FALSE(circle_shape.inside(0.1, 0.1)); // Outside
    EXPECT_TRUE(circle_shape.inside(0.9, 0.5));  // On edge
}

// Test: Point inside without loading shape
TEST_F(ColliderTest, Collider_PointInsideWithoutLoading) {
    collider shape;
    EXPECT_FALSE(shape.inside(0.5, 0.5)); // Should handle not loaded case
}

// Test: Reverse winding order shape
TEST_F(ColliderTest, Collider_ReverseWindingOrder) {
    collider reversed_shape;
    reversed_shape.load_shape(reversed_square_shape_file);
    EXPECT_TRUE(reversed_shape.is_loaded());
    EXPECT_EQ(reversed_shape.get_vertices().size(), 4);
    EXPECT_TRUE(reversed_shape.inside(0.5, 0.5));  // Inside
    EXPECT_FALSE(reversed_shape.inside(1.5, 0.5)); // Outside
}

File: tests/python/test_collider.py
"""
Python tests for the Collider vessel-wall boundary tester.

Mirrors the C++ GoogleTest suite in tests/test_collider.cpp.
Synthetic shapes are created in a temporary directory; the TCABR wall file
is loaded from tests/data/.
"""

import math
import os
import tempfile

import numpy as np
import pytest
import maglib
from maglib import Collider

DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "data")


# ── Fixtures ──────────────────────────────────────────────────────────────────

@pytest.fixture
def square_file(tmp_path):
    """Unit square (0,0)–(1,1), counter-clockwise."""
    p = tmp_path / "square.txt"
    p.write_text("0.0 0.0\n1.0 0.0\n1.0 1.0\n0.0 1.0\n")
    return str(p)


@pytest.fixture
def octagon_file(tmp_path):
    """Approximate circle: regular octagon centred at (0.5,0.5), radius 0.4."""
    cx, cz, r = 0.5, 0.5, 0.4
    lines = "\n".join(
        f"{cx + r * math.cos(2 * math.pi * i / 8):.15f} "
        f"{cz + r * math.sin(2 * math.pi * i / 8):.15f}"
        for i in range(8)
    )
    p = tmp_path / "octagon.txt"
    p.write_text(lines + "\n")
    return str(p)


@pytest.fixture
def invalid_file(tmp_path):
    """Only two vertices — too few for a polygon."""
    p = tmp_path / "invalid.txt"
    p.write_text("0.0 0.0\n1.0 0.0\n")
    return str(p)


@pytest.fixture
def reversed_square_file(tmp_path):
    """Unit square with clockwise (reversed) winding order."""
    p = tmp_path / "reversed.txt"
    p.write_text("0.0 1.0\n1.0 1.0\n1.0 0.0\n0.0 0.0\n")
    return str(p)


@pytest.fixture
def tcabr_file():
    return os.path.join(DATA_DIR, "tcabr_first_wall.txt")


# ── Construction ──────────────────────────────────────────────────────────────

def test_default_construction():
    c = Collider()
    assert not c.is_loaded()


# ── Loading ───────────────────────────────────────────────────────────────────

def test_load_square(square_file):
    c = Collider()
    assert c.load_shape(square_file)
    assert c.is_loaded()
    assert len(c.get_vertices()) == 4


def test_load_octagon(octagon_file):
    c = Collider()
    assert c.load_shape(octagon_file)
    assert c.is_loaded()
    assert len(c.get_vertices()) == 8


def test_load_invalid(invalid_file):
    c = Collider()
    assert not c.load_shape(invalid_file)


def test_load_nonexistent():
    c = Collider()
    assert not c.load_shape("/nonexistent/path.txt")


def test_load_tcabr(tcabr_file):
    c = Collider()
    assert c.load_shape(tcabr_file)
    assert c.is_loaded()
    assert len(c.get_vertices()) == 68


# ── Inside/outside tests ──────────────────────────────────────────────────────

def test_square_inside(square_file):
    c = Collider()
    c.load_shape(square_file)
    assert c.inside(0.5, 0.5)   # centre
    assert not c.inside(1.5, 0.5)  # clearly outside
    assert c.inside(0.0, 0.5)   # on left edge


def test_octagon_inside(octagon_file):
    c = Collider()
    c.load_shape(octagon_file)
    assert c.inside(0.5, 0.5)   # centre
    assert not c.inside(0.1, 0.1)  # outside


def test_inside_without_loading():
    c = Collider()
    assert not c.inside(0.5, 0.5)


def test_reversed_winding_square(reversed_square_file):
    c = Collider()
    assert c.load_shape(reversed_square_file)
    assert c.inside(0.5, 0.5)
    assert not c.inside(1.5, 0.5)


# ── get_vertices shape ────────────────────────────────────────────────────────

def test_get_vertices_returns_numpy_array(square_file):
    c = Collider()
    c.load_shape(square_file)
    verts = c.get_vertices()
    assert isinstance(verts, np.ndarray)
    assert verts.shape == (4, 2)
    assert verts.dtype == np.float64


def test_get_vertices_values(square_file):
    c = Collider()
    c.load_shape(square_file)
    verts = c.get_vertices()
    expected = np.array([[0.0, 0.0], [1.0, 0.0], [1.0, 1.0], [0.0, 1.0]])
    np.testing.assert_array_equal(verts, expected)
